pkg/session/session.go

```diff
@@ -40,18 +42,60 @@ import (
 	"github.com/vmware/govmomi/vim25/methods"
 	"github.com/vmware/govmomi/vim25/soap"
 	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
 
 	infrav1 "sigs.k8s.io/cluster-api-provider-vsphere/apis/v1beta1"
 	"sigs.k8s.io/cluster-api-provider-vsphere/pkg/constants"
 )
 
+const (
+	metricNameSpace        = "session"
+	metricLabelServer      = "server"
+	metricLabelDC          = "dc"
+	metricLabelUsername    = "username"
+	metricLabelSessionKey  = "sessionKey"
+	metricLabelSessionKeys = "sessionKeys"
+)
+
 var (
 	// global Session map against sessionKeys in map[sessionKey]Session.
 	sessionCache sync.Map
 
 	// mutex to control access to the GetOrCreate function to avoid duplicate
 	// session creations on startup.
 	sessionMU sync.Mutex
+
+	sessionCacheMetric = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: metricNameSpace,
+			Name:      "cached_num",
+		},
+		[]string{metricLabelSessionKeys},
+	)
+
+	getSessionMetric = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: metricNameSpace,
+			Name:      "get",
+		},
+		[]string{metricLabelServer, metricLabelDC, metricLabelUsername},
+	)
+
+	createSessionMetric = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: metricNameSpace,
+			Name:      "create",
+		},
+		[]string{metricLabelServer, metricLabelDC, metricLabelUsername},
+	)
+
+	deleteSessionMetric = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: metricNameSpace,
+			Name:      "delete",
+		},
+		[]string{metricLabelSessionKey},
+	)
```

Comment:
I'll defer this to @sbueringer and @chrischdi who have more expertise than me on metrics, but I think we should consider having only one metric about VSphere session operations, with get/create/delete (the operation) as a label

Reply:
Thanks for this, That is definitely a good option, and it also makes sense to me.

Reply:
💯 let's use the different verbs label instead having a metric per verb.